lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -566,27 +566,62 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
   return tryToConcatenateStrings(AI, Builder);
 }
 
+// TODO:
+// Move this concrete type propagation for existentials out of sil-combine
+// We can then use DominanceAnalysis which is the right thing to do here
+// We avoid using it here due to compilation speed implications
+static bool fastDominates(SILInstruction *a, SILInstruction *b) {
+  auto aBlock = a->getParent(), bBlock = b->getParent();
+
+  // If the blocks are different, bail
+  if (aBlock != bBlock)
+    return true;
+
+  // Otherwise, they're in the same block, and we just need to check
+  // whether B comes after A.  This is a non-strict computation.
+  auto aIter = a->getIterator();
+  auto bIter = b->getIterator();
+  auto fIter = aBlock->begin();
+  while (bIter != fIter) {
```

Comment:
This look potentially introduces a compile time problem (and yes, we call this code in other places, too). So at least you should check if it does not result in quadratic behavior, e.g. in case of large basic blocks

Reply:
After the cleanups I will test compile time performance using Graydon's new CI option as a first step.